### PR TITLE
Add option to specify container restart policy

### LINKED ISSUE #578: Specify container restart policy
**Overview of feature request**
Radius users should be able to specify the behavior of a stopped container in the restartPolicy property, choosing between never, always, or onfailure.

```bicep
  resource nodeapplication 'Components' = {
    name: 'nodeapp'
    kind: 'radius.dev/Container@v1alpha1'
    properties: {
      run: {
        container: {
          image: 'radiusteam/tutorial-nodeapp'
        }
        restartPolicy: 'always'
        ...
```

## Estimation (number of half days)

2

[AB#4126](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/4126)

### PR DESCRIPTION
**Pull request by shubham1172**
# Description

This PR adds an option to specify the container restart policy.

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #578

### Summary
🚀📝🧪

This pull request adds a new feature to the `radius` project that allows users to specify the restart policy for containers in the `v20231001preview` API version. It updates the typespec file, the OpenAPI specification, the data model, the API model, the rendering logic, and the tests to support the new `RestartPolicy` property and enum type for containers.

> _`restartPolicy`_
> _New feature for containers_
> _Autumn of failures_

### Walkthrough
*  Add the `RestartPolicy` property to the `ContainerProperties` and `ContainerResourceUpdateProperties` types in the data model and the API model ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-21034b9d798fa570ecae4b7d54142c7f94d9d0f1b39c5f45f0c4c859674b19fbR358-R360), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-21034b9d798fa570ecae4b7d54142c7f94d9d0f1b39c5f45f0c4c859674b19fbR439-R441), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-5275155acd7f7a59f0301cfc95f8dd38f72f22795e3197c6a150bae13dd5d5f3R81), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R3121-R3124), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R3263-R3266), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-2f8e96d0546ae7b99ee68e516a354825696d77ca06e49bb6f82ab1caaa778522R71-R73))
*  Add the `RestartPolicy` enum type to the data model and the API model, with three possible values: `Always`, `OnFailure`, and `Never` ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-07cd8284519deaa5e625f1fe393cb48bdd1cc621c9a4656ca43cc6049dfe7962R307-R327), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R4865-R4894), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-2f8e96d0546ae7b99ee68e516a354825696d77ca06e49bb6f82ab1caaa778522R87-R98))
*  Add the conversion logic for the `RestartPolicy` property between the data model and the API model, using helper functions to map the enum type to the string type and vice versa ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-c019a02a8a768b2eec23949592ff1e7b5a167de082059261b3c3372e2870dacaR140), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-c019a02a8a768b2eec23949592ff1e7b5a167de082059261b3c3372e2870dacaR269), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-c019a02a8a768b2eec23949592ff1e7b5a167de082059261b3c3372e2870dacaR601-R630))
*  Add the JSON serialization and deserialization logic for the `RestartPolicy` property, using the `populate` and `unpopulate` helper functions to handle the JSON keys and values ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-bc1502ced9e29d6d61dc028d22603e863f59a981bbc54a59879ba3177d8e62e9R766), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-bc1502ced9e29d6d61dc028d22603e863f59a981bbc54a59879ba3177d8e62e9R808-R810), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-bc1502ced9e29d6d61dc028d22603e863f59a981bbc54a59879ba3177d8e62e9R949), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-bc1502ced9e29d6d61dc028d22603e863f59a981bbc54a59879ba3177d8e62e9R987-R989))
*  Add the rendering logic for the `RestartPolicy` property, setting the `RestartPolicy` field of the Kubernetes deployment resource to the value of the property, if not empty ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-3da981bf20df07918a646eafcd867d6c95abac1f3aef1ed1b50a29f3f4cc8636R645-R649))
*  Add the test case for the rendering logic, verifying that the rendered deployment has the same restart policy as the resource ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-6f1219f263ab06a1493ac76960e2ab8cbe647e83e926bff7d13988f393334f94R1266-R1297))
*  Add the documentation for the `RestartPolicy` property and the `RestartPolicy` enum type in the OpenAPI specification and the typespec file, using doc comments and descriptions ([link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R3121-R3124), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R3263-R3266), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-ac35cdfa65680d5d2be2d0513a686f4651b9a41eb951d05dcd446bc927090d68R4865-R4894), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-2f8e96d0546ae7b99ee68e516a354825696d77ca06e49bb6f82ab1caaa778522R71-R73), [link](https://github.com/radius-project/radius/pull/6546/files?diff=unified&w=0#diff-2f8e96d0546ae7b99ee68e516a354825696d77ca06e49bb6f82ab1caaa778522R87-R98))

### CODE REVIEW

**Comment by rynowak** [comment]
@ytimocin - can you help review this?

**Review by ytimocin — COMMENTED** [review]
Thank you for this addition and it looks great so far. We may need to add some unit tests for the container converter. After that addition I can approve.

File: pkg/corerp/api/v20231001preview/container_conversion.go [hunk 1]
```diff
@@ -265,6 +266,7 @@ func (dst *ContainerResource) ConvertFrom(src v1.DataModelInterface) error {
 		Runtimes:             fromRuntimePropertiesDataModel(c.Properties.Runtimes),
 		Resources:            fromResourceReferencesDataModel(c.Properties.Resources),
 		ResourceProvisioning: fromContainerResourceProvisioningDataModel(c.Properties.ResourceProvisioning),
+		RestartPolicy:        fromRestartPolicyDataModel(c.Properties.RestartPolicy),
```
[the remark this answers is not in the document]
Comment (shubham1172) [reply pkg/corerp/api/v20231001preview/container_conversion.go:269]: Thanks @ytimocin, added the tests!

**Review by lakshmimsft — APPROVED** [review]

**Review by rynowak — APPROVED** [review]

**Comment by rynowak** [comment]
@shalabhms - can you help mark sure this PR gets merged? It's approved at this point but we'll have to bypass some of the checks because of failing automation.

**Review by ytimocin — APPROVED** [review]

**Comment by shalabhms** [comment]
Thanks @shubham1172 for your contribution!

**Pull request merged**